### PR TITLE
[quant][pt2e] Support int16 quantization

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108453

Summary:
Previously we can only use native pytorch int dtypes that has corresponding quantized dtypes (e.g. quint8, qint8), this
PR removes this assumption in observers/fake_quants so that users can use all pytorch native dtypes (except for int64, we can add it later if need)
the main addition here is int16.

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/ao/quantization/observer.py [hunk 1]
```diff
@@ -330,7 +338,7 @@ def _calculate_qparams(
             max_val_pos = torch.max(-min_val_neg, max_val_pos)
             scale = max_val_pos / (float(quant_max - quant_min) / 2)
             scale = torch.max(scale, self.eps)
-            if self.dtype == torch.quint8:
+            if self.dtype in [torch.quint8, torch.uint8]:
```
Comment (kimishpatel) [inline torch/ao/quantization/observer.py:341]: why did this have to be changed?

File: test/quantization/pt2e/test_quantize_pt2e.py [hunk 2]
```diff
@@ -1277,6 +1281,66 @@ def validate(self, model: torch.fx.GraphModule) -> None:
             m, expected_node_list=node_list, expected_node_occurrence=node_occurrence
         )
 
+    def test_int16(self):
+        class Int16ActQuantizer(Quantizer):
+            def annotate(self, model: torch.fx.GraphModule) -> torch.fx.GraphModule:
+                # using int32 to simulate int16
+                int16_qspec = QuantizationSpec(
+                    dtype=torch.int32,
```
Comment (kimishpatel) [inline test/quantization/pt2e/test_quantize_pt2e.py:1289]: this says dtype=int32 not int16

Comment (kimishpatel) [reply test/quantization/pt2e/test_quantize_pt2e.py:1289]: oh nevermind, I see the restriction on the values. I thought int16 is supported torch dtype but not uint16? Then why not use that dtype directly

**Review by kimishpatel — CHANGES_REQUESTED** [review]
sending back for test fix.

For dtype change in fx workflow, previously it was working because we were mapping torch dtypes to quantized dtypes?

File: test/quantization/pt2e/test_quantize_pt2e.py (hunk 2 above)
Comment (jerryzh168) [reply test/quantization/pt2e/test_quantize_pt2e.py:1289]: oh I think I can use torch.int16 here directly, not sure why I used int32 in the beginning

File: torch/ao/quantization/observer.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/observer.py:341]: this is to make sure the code works for both quint8 and uint8 dtypes

File: test/quantization/pt2e/test_quantize_pt2e.py [hunk 3]
```diff
@@ -2114,7 +2177,7 @@ def _get_getitem_nodes(m: torch.fx.GraphModule):
 
     def test_qat_update_shared_qspec(self):
         """
-        Test the case where nodes used in SharedQuantizationSpec were replaced
+        Test the case where nodes used in SharedQuantizationpec were replaced
```
Comment (andrewor14) [inline test/quantization/pt2e/test_quantize_pt2e.py:2180]: typo

**Comment by jerryzh168** [comment]
> For dtype change in fx workflow, previously it was working because we were mapping torch dtypes to quantized dtypes?

yeah that's correct, we are not really using torch dtypes before in observer or fx flows

**Review by kimishpatel — APPROVED** [review]

**Pull request closed without merging**